### PR TITLE
Add  Dict Obs support for DQN

### PR DESCRIPTION
**Pull request by suijth**
## Description
Add MultiInputPolicy for DQN.

Tried to make minimal changes to support this. Did separate out extractors like sb3's MultiInputPolicy  

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [X] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [X] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [X] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [X] I have reformatted the code using `make format` (**required**)
- [X] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [X] I have ensured `make pytest` and `make type` both pass. (**required**)
- [X] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

**Comment by araffin** [comment]
Hello,
thanks for the PR.
When you check items in the checklist (especially all the `make' commands), it means you have executed them, not that you will let the CI do the job.

Please avoid any `type: ignore` unless necessary too.

**Comment by suijth** [comment]
> Hello, thanks for the PR. When you check items in the checklist (especially all the `make' commands), it means you have executed them, not that you will let the CI do the job.

My bad. I was using ubuntu's [mypy](https://manpages.ubuntu.com/manpages/lunar/man1/mypy.1.html) which was not catching these errors not sure why. I'll probably create a new branch without these commits, it'll be cleaner. 

> Please avoid any `type: ignore` unless necessary too.

Noted. Removed them at most places.

File: sbx/common/policies.py [hunk 1]
```diff
@@ -82,16 +94,22 @@ def predict(
         return actions, state
 
     def prepare_obs(self, observation: Union[np.ndarray, dict[str, np.ndarray]]) -> tuple[np.ndarray, bool]:
+        return self._prepare_obs(observation, self.observation_space)
```
Comment (araffin) [inline sbx/common/policies.py:97]: why do you need another method?

File: sbx/dqn/policies.py [hunk 2]
```diff
@@ -148,3 +152,86 @@ def build(self, key: jax.Array, lr_schedule: Schedule) -> jax.Array:
...
+
+class MultiInputQNetwork(nn.Module):
+    observation_space: spaces.Space
+    n_actions: int
+    cnn_output_dim: int = 256
+    n_units: int = 256
+    activation_fn: Callable[[jnp.ndarray], jnp.ndarray] = nn.relu
+
+    def setup(self):
+        def layer(subspace):
```
Comment (araffin) [inline sbx/dqn/policies.py:165]: does this have to be in `setup()`, `layer()` cannot be put at the root or in the utils?
also missing type annotation

File: sbx/dqn/policies.py [hunk 3]
```diff
@@ -148,3 +152,86 @@ def build(self, key: jax.Array, lr_schedule: Schedule) -> jax.Array:
...
+                )
+            elif isinstance(subspace, spaces.Discrete):
+                return OneHot(num_classes=subspace.n)
+            return Flatten()
+
+        self.extractors = jax.tree_map(layer, self.observation_space.spaces)
+
+    @nn.compact
+    def __call__(self, observations: dict[str, jnp.ndarray]) -> jnp.ndarray:
+        encoded_tensors = jax.tree_map(lambda extractor, x: extractor(x), self.extractors, flax.core.freeze(observations))
```
Comment (araffin) [inline sbx/dqn/policies.py:180]: `flax.core.freeze()` is needed otherwise jax complains?

File: sbx/common/policies.py (hunk 1 above)
Comment (suijth) [reply sbx/common/policies.py:101]: Added a comment

File: sbx/dqn/policies.py (hunk 3 above)
Comment (suijth) [reply sbx/dqn/policies.py:169]: yes, it is storing extractors as a frozen dict so we need to match the type to use jax.tree_map.
other way around would be to unfreeze extractors and convert them to OrderedDict(which is the type of observations here)

File: sbx/dqn/policies.py (hunk 2 above)
Comment (suijth) [reply sbx/dqn/policies.py]: did not find a appropriate util file. Move the method to the root.

File: sbx/dqn/policies.py [hunk 4]
```diff
@@ -148,3 +152,86 @@ def build(self, key: jax.Array, lr_schedule: Schedule) -> jax.Array:
...
+                learning_rate=lr_schedule(1),  # type: ignore[call-arg]
+                **self.optimizer_kwargs,
+            ),
+        )
+
+        self.qf.apply = jax.jit(self.qf.apply)  # type: ignore[method-assign]
+
+        return key
+
+    def prepare_obs(  # type: ignore[override]
```
Comment (araffin) [inline sbx/dqn/policies.py:220]: this preprocessing is done only for the data collection, right?

but it should not be needed because the preprocessing is done inside `__call__` by the multi input q network?


I started rewriting SB3 `preprocessing.py` and then realized it should not be needed.
```python
from typing import Union

import jax
import jax.numpy as jnp
from gymnasium import spaces
from stable_baselines3.common.preprocessing import is_image_space


def preprocess_obs(
    obs: Union[jax.Array, dict[str, jax.Array]],
    observation_space: spaces.Space,
    normalize_images: bool = True,
) -> Union[jax.Array, dict[str, jax.Array]]:
    """
    Preprocess observation to be to a neural network.
    For images, it normalizes the values by dividing them by 255 (to have values in [0, 1])
    For discrete observations, it create a one hot vector.

    :param obs: Observation
    :param observation_space:
    :param normalize_images: Whether to normalize images or not
        (True by default)
    :return:
    """
    if isinstance(observation_space, spaces.Dict):
        # Do not modify by reference the original observation
        assert isinstance(obs, dict), f"Expected dict, got {type(obs)}"
        preprocessed_obs = jax.tree.map(
            lambda observation, obs_space: preprocess_obs(observation, obs_space, normalize_images=normalize_images),
            obs,
            observation_space.spaces,
        )
        return preprocessed_obs  # type: ignore[return-value]

    assert not isinstance(obs, dict)

    if isinstance(observation_space, spaces.Box):
        if normalize_images and is_image_space(observation_space):
            return obs.astype(jnp.float32) / 255.0
        return obs.astype(jnp.float32)

    elif isinstance(observation_space, spaces.Discrete):
        # One hot encoding and convert to float to avoid errors
        return jax.nn.one_hot(obs, num_classes=int(observation_space.n)).astype(jnp.float32)

    elif isinstance(observation_space, spaces.MultiBinary):
        return obs.astype(jnp.float32)
    else:
        # MISSING: spaces.MultiDiscrete
        raise NotImplementedError(f"Preprocessing not implemented for {observation_space}")
```

Comment (suijth) [reply sbx/dqn/policies.py:220]: > this preprocessing is done only for the data collection, right?     

yes

> but it should not be needed because the preprocessing is done inside __call__ by the multi input q network?

yes, part of it is done inside `__call__` (pytorch image -> jax transforming and normalising).  Inside prepare_obs we are converting non vectorized obs to vectorized and image obs to pytorch format. 
converting image to pytorch then to jax format might be redundant, we can find a way to avoid it 
but we'll need the  vectorizing part thus we need this prepare_obs. 

This is my understanding.
